### PR TITLE
[test] fix TestLoadBlockchainfromDB

### LINKED ISSUE #3520: [Flaky Test] - TestLoadBlockchainfromDB
https://github.com/iotexproject/iotex-core/runs/7293611023?check_suite_focus=true

'''
=== CONT  TestLoadBlockchainfromDB
    integrity_test.go:1146: 
        	Error Trace:	integrity_test.go:1146
        	            				integrity_test.go:1336
        	Error:      	Not equal: 
        	            	expected: 24
        	            	actual  : 16
        	Test:       	TestLoadBlockchainfromDB
'''

### PR DESCRIPTION
**Pull request by millken**
# Description
Fixed `make test` sometimes failed

Fixes #3520 

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: blockchain/integrity/integrity_test.go [hunk 1]
```diff
@@ -832,22 +832,17 @@ func TestBlockchain_MintNewBlock_PopAccount(t *testing.T) {
...
-	mu      sync.RWMutex
+	counter int32
 }
 
 func (ms *MockSubscriber) ReceiveBlock(blk *block.Block) error {
-	ms.mu.Lock()
 	tsfs, _ := classifyActions(blk.Actions)
-	ms.counter += len(tsfs)
-	ms.mu.Unlock()
+	atomic.AddInt32(&ms.counter, int32(len(tsfs)))
```
Comment (dustinxie) [inline blockchain/integrity/integrity_test.go:840]: good fix

File: blockchain/pubsubmanager.go [hunk 2]
```diff
@@ -93,9 +93,10 @@ func (ps *pubSub) RemoveBlockListener(s BlockCreationSubscriber) error {
 
 // SendBlockToSubscribers sends block to every subscriber by using buffer channel
 func (ps *pubSub) SendBlockToSubscribers(blk *block.Block) {
-	ps.lock.Lock()
-	defer ps.lock.Unlock()
-	for _, elem := range ps.blocklisteners {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+	blocklisteners := ps.blocklisteners
+	for _, elem := range blocklisteners {
```
Comment (dustinxie) [inline blockchain/pubsubmanager.go:99]: this change is not necessary?

Comment (millken) [reply blockchain/pubsubmanager.go:99]: `RLock`,`RUnlock` is better here

Comment (dustinxie) [reply blockchain/pubsubmanager.go:99]: i mean, no need to define `blocklisteners := ps.blocklisteners`, use like before
```
for _, elem := range ps.blocklisteners
```

Comment (millken) [reply blockchain/pubsubmanager.go]: done

**Review by dustinxie — APPROVED** [review]

File: blockchain/integrity/integrity_test.go [hunk 3]
```diff
@@ -1142,6 +1137,8 @@ func TestLoadBlockchainfromDB(t *testing.T) {
 		height := bc.TipHeight()
 		fmt.Printf("Open blockchain pass, height = %d\n", height)
 		require.NoError(addTestingTsfBlocks(cfg, bc, dao, ap))
+		//make sure pubsub is completed
+		time.Sleep(time.Millisecond * 200)
```
Comment (Liuhaai) [inline blockchain/integrity/integrity_test.go:1141]: use waitutil instead of sleep in the test

Comment (millken) [reply blockchain/integrity/integrity_test.go]: done, thanks

File: blockchain/integrity/integrity_test.go [hunk 4]
```diff
@@ -1142,7 +1137,12 @@ func TestLoadBlockchainfromDB(t *testing.T) {
 		height := bc.TipHeight()
 		fmt.Printf("Open blockchain pass, height = %d\n", height)
 		require.NoError(addTestingTsfBlocks(cfg, bc, dao, ap))
-		require.NoError(bc.Stop(ctx))
+		//make sure pubsub is completed
+		err = testutil.WaitUntil(200*time.Millisecond, 3*time.Second, func() (bool, error) {
```
Comment (Liuhaai) [inline blockchain/integrity/integrity_test.go:1141]:
```
err = testutil.WaitUntil(200*time.Millisecond, 3*time.Second, func() (bool, error) {
			return ms.Counter() == 24, nil
		})
		require.NoError(err)
		require.NoError(bc.Stop(ctx))
```

Comment (millken) [reply blockchain/integrity/integrity_test.go:1141]: Your fix looks better

**Pull request merged**